File: examples/node/basic_usage.js
#!/usr/bin/env node
/**
 * StateSet iCommerce - Node.js Example
 *
 * Demonstrates the full commerce workflow:
 * - Customer creation
 * - Product catalog management
 * - Inventory tracking
 * - Order processing
 * - Analytics
 *
 * Run with: node basic_usage.js
 */

const { Commerce } = require('@stateset/embedded');

async function main() {
  console.log('=== StateSet iCommerce - Node.js Example ===\n');

  // Initialize commerce with in-memory database
  const commerce = new Commerce(':memory:');
  console.log('✓ Commerce initialized\n');

  // 1. Create a customer
  console.log('1. Creating customer...');
  const customer = await commerce.customers.create({
    email: 'alice@example.com',
    firstName: 'Alice',
    lastName: 'Smith',
    phone: '+1-555-0123'
  });
  console.log(`   Created customer: ${customer.firstName} ${customer.lastName} (${customer.email})`);

  // 2. Create products
  console.log('\n2. Creating products...');
  const widget = await commerce.products.create({
    name: 'Premium Widget',
    sku: 'WIDGET-001',
    price: 29.99,
    description: 'A high-quality widget for all your needs'
  });
  console.log(`   Created product: ${widget.name} (${widget.slug})`);

  const gadget = await commerce.products.create({
    name: 'Super Gadget',
    sku: 'GADGET-001',
    price: 49.99,
    description: 'An amazing gadget'
  });
  console.log(`   Created product: ${gadget.name} (${gadget.slug})`);

  // 3. Create inventory
  console.log('\n3. Setting up inventory...');
  await commerce.inventory.createItem({
    sku: 'WIDGET-001',
    name: 'Premium Widget',
    initialQuantity: 100
  });
  console.log('   Created inventory for WIDGET-001 (100 units)');

  await commerce.inventory.createItem({
    sku: 'GADGET-001',
    name: 'Super Gadget',
    initialQuantity: 50
  });
  console.log('   Created inventory for GADGET-001 (50 units)');

  // Check stock
  const widgetStock = await commerce.inventory.getStock('WIDGET-001');
  console.log(`   Stock check WIDGET-001: ${widgetStock.totalAvailable} available`);

  // 4. Create an order
  console.log('\n4. Creating order...');
  const order = await commerce.orders.create({
    customerId: customer.id,
    items: [
      {
        productId: widget.id,
        sku: 'WIDGET-001',
        name: 'Premium Widget',
        quantity: 2,
        unitPrice: 29.99
      },
      {
        productId: gadget.id,
        sku: 'GADGET-001',
        name: 'Super Gadget',
        quantity: 1,
        unitPrice: 49.99
      }
    ],
    currency: 'USD'
  });
  console.log(`   Created order ${order.orderNumber} (total: $${order.totalAmount})`);

  // 5. Process the order
  console.log('\n5. Processing order...');

  // Update order status
  let updatedOrder = await commerce.orders.updateStatus(order.id, 'confirmed');
  console.log(`   Order status: ${updatedOrder.status}`);

  // Adjust inventory (fulfill)
  await commerce.inventory.adjust('WIDGET-001', -2, 'Order fulfillment');
  await commerce.inventory.adjust('GADGET-001', -1, 'Order fulfillment');
  console.log('   Inventory adjusted');

  // Ship the order
  updatedOrder = await commerce.orders.ship(order.id, 'TRACK123456');
  console.log(`   Order shipped with tracking: ${updatedOrder.trackingNumber}`);

  // 6. Check final inventory
  console.log('\n6. Final inventory check...');
  const finalWidgetStock = await commerce.inventory.getStock('WIDGET-001');
  console.log(`   WIDGET-001: ${finalWidgetStock.totalAvailable} available (was 100)`);

  const finalGadgetStock = await commerce.inventory.getStock('GADGET-001');
  console.log(`   GADGET-001: ${finalGadgetStock.totalAvailable} available (was 50)`);

  // 7. Analytics
  console.log('\n7. Analytics...');
  const salesSummary = await commerce.analytics.getSalesSummary('today');
  console.log(`   Revenue: $${salesSummary.totalRevenue}`);
  console.log(`   Orders: ${salesSummary.orderCount}`);
  console.log(`   AOV: $${salesSummary.averageOrderValue}`);

  // 8. Summary
  console.log('\n=== Summary ===');
  const customers = await commerce.customers.list();
  const products = await commerce.products.list();
  const orders = await commerce.orders.list();
  console.log(`Customers: ${customers.length}`);
  console.log(`Products: ${products.length}`);
  console.log(`Orders: ${orders.length}`);

  console.log('\n✓ Example completed successfully!');
}

main().catch(console.error);

File: examples/python/basic_usage.py
#!/usr/bin/env python3
"""
StateSet iCommerce - Python Example

Demonstrates the full commerce workflow:
- Customer creation
- Product catalog management
- Inventory tracking
- Order processing
- Analytics

Run with: python basic_usage.py
"""

from stateset import Commerce


def main():
    print("=== StateSet iCommerce - Python Example ===\n")

    # Initialize commerce with in-memory database
    commerce = Commerce(":memory:")
    print("✓ Commerce initialized\n")

    # 1. Create a customer
    print("1. Creating customer...")
    customer = commerce.customers.create(
        email="alice@example.com",
        first_name="Alice",
        last_name="Smith",
        phone="+1-555-0123"
    )
    print(f"   Created customer: {customer.first_name} {customer.last_name} ({customer.email})")

    # 2. Create products
    print("\n2. Creating products...")
    widget = commerce.products.create(
        name="Premium Widget",
        sku="WIDGET-001",
        price=29.99,
        description="A high-quality widget for all your needs"
    )
    print(f"   Created product: {widget.name} ({widget.slug})")

    gadget = commerce.products.create(
        name="Super Gadget",
        sku="GADGET-001",
        price=49.99,
        description="An amazing gadget"
    )
    print(f"   Created product: {gadget.name} ({gadget.slug})")

    # 3. Create inventory
    print("\n3. Setting up inventory...")
    commerce.inventory.create_item(
        sku="WIDGET-001",
        name="Premium Widget",
        initial_quantity=100
    )
    print("   Created inventory for WIDGET-001 (100 units)")

    commerce.inventory.create_item(
        sku="GADGET-001",
        name="Super Gadget",
        initial_quantity=50
    )
    print("   Created inventory for GADGET-001 (50 units)")

    # Check stock
    widget_stock = commerce.inventory.get_stock("WIDGET-001")
    print(f"   Stock check WIDGET-001: {widget_stock.total_available} available")

    # 4. Create an order
    print("\n4. Creating order...")
    order = commerce.orders.create(
        customer_id=customer.id,
        items=[
            {
                "product_id": widget.id,
                "sku": "WIDGET-001",
                "name": "Premium Widget",
                "quantity": 2,
                "unit_price": "29.99"
            },
            {
                "product_id": gadget.id,
                "sku": "GADGET-001",
                "name": "Super Gadget",
                "quantity": 1,
                "unit_price": "49.99"
            }
        ],
        currency="USD"
    )
    print(f"   Created order {order.order_number} (total: ${order.total_amount})")

    # 5. Process the order
    print("\n5. Processing order...")

    # Update order status
    order = commerce.orders.update_status(order.id, "confirmed")
    print(f"   Order status: {order.status}")

    # Adjust inventory (fulfill)
    commerce.inventory.adjust("WIDGET-001", -2, "Order fulfillment")
    commerce.inventory.adjust("GADGET-001", -1, "Order fulfillment")
    print("   Inventory adjusted")

    # Ship the order
    order = commerce.orders.ship(order.id, "TRACK123456")
    print(f"   Order shipped with tracking: {order.tracking_number}")

    # 6. Check final inventory
    print("\n6. Final inventory check...")
    final_widget_stock = commerce.inventory.get_stock("WIDGET-001")
    print(f"   WIDGET-001: {final_widget_stock.total_available} available (was 100)")

    final_gadget_stock = commerce.inventory.get_stock("GADGET-001")
    print(f"   GADGET-001: {final_gadget_stock.total_available} available (was 50)")

    # 7. Analytics
    print("\n7. Analytics...")
    sales_summary = commerce.analytics.sales_summary(period="today")
    print(f"   Revenue: ${sales_summary.total_revenue}")
    print(f"   Orders: {sales_summary.order_count}")
    print(f"   AOV: ${sales_summary.average_order_value}")

    # Get top products
    top_products = commerce.analytics.top_products(limit=5)
    print(f"   Top products: {len(top_products)}")

    # 8. Demand forecasting
    print("\n8. Demand forecasting...")
    forecasts = commerce.analytics.demand_forecast(days_ahead=30)
    for forecast in forecasts:
        if forecast.days_until_stockout and forecast.days_until_stockout < 14:
            print(f"   WARNING: {forecast.sku} will stock out in {forecast.days_until_stockout} days")
        else:
            print(f"   {forecast.sku}: {forecast.predicted_demand} units predicted")

    # 9. Summary
    print("\n=== Summary ===")
    customers_list = commerce.customers.list()
    products_list = commerce.products.list()
    orders_list = commerce.orders.list()
    print(f"Customers: {len(customers_list)}")
    print(f"Products: {len(products_list)}")
    print(f"Orders: {len(orders_list)}")

    print("\n✓ Example completed successfully!")


if __name__ == "__main__":
    main()
